feat: change CLI to dynamically generate args from flavor options

mikeseese opened the pull request:
This PR is part of #689 which changes `@ganache/ethereum-options` to include the necessary values/types to support the new CLI `src/args.ts` file which dynamically generates the associated `yargs.option()` calls

This PR also resolves #690

mikeseese committed:
separate options from flavor

takes the commits from PR #702 and applies non-filecoin contents

davidmurdoch committed:
make create.ts use the original npm command's argv (#707)

mikeseese committed:
dynamically generate args for cli using flavors and options

mikeseese committed:
get cli running with new dynamic args

mikeseese committed:
run link-ts-references

mikeseese commented on package.json:
```diff
@@ -8,7 +8,7 @@
   "scripts": {
     "build": "lerna run build && shx cp -r ./src/packages/ganache/dist/web ./docs/assets/js/ganache",
-    "clean": "npx shx rm -rf node_modules && npx lerna clean -y && npm run tsc.clean",
+    "clean": "npm run tsc.clean && npx lerna clean -y && npx shx rm -rf node_modules",
```
I had to reverse these because the first command `npx shx rm -rf node_modules` removes a non-global installation of lerna, causing the following commands to fail

davidmurdoch replied:
lerna is run with `npx` here, so that shouldn't be a problem. There might be another issue with the way ganache is configured. hmmm

mikeseese replied:
All I know is running `npm run clean` would finally get to `npm run tsc.clean` and I would get `lerna not found`. I'm guessing this has been working for you because you have a globally installed version of lerna, which isn't a requirement of this repo.

After changing the order of this command, I can run `npm run clean` and choosing to not look into the issue further and call it done. I don't think this is a configuration issue, nor do I think I should have to reinstall lerna (because that's what npx will do if I don't have it) just to finish running the clean command lol

davidmurdoch replied:
I'm good with the change; it make sense... I just found it strange.


Only commenting this just in case we need to come back to this one day:

I don't have lerna installed globally, and `npm run clean` (`npx shx rm -rf node_modules && npx lerna clean -y && npm run tsc.clean`) looks to be working on my machine.


```console
david at david-kubuntu in ~/work/ganache-core on chore/separate-options-from-flavors-next [!$]
$ lerna
lerna: command not found

david at david-kubuntu in ~/work/ganache-core on chore/separate-options-from-flavors-next [!$]
$ npx shx rm -rf node_modules

david at david-kubuntu in ~/work/ganache-core on chore/separate-options-from-flavors-next [!$]
$ npm run clean

> root@ clean /home/david/work/ganache-core
> npx shx rm -rf node_modules && npx lerna clean -y && npm run tsc.clean

npx: installed 21 in 3.73s
npx: installed 766 in 34.004s
lerna notice cli v3.22.1
lerna info versioning independent
lerna info clean removing /home/david/work/ganache-core/src/chains/ethereum/ethereum/node_modules
lerna info clean removing /home/david/work/ganache-core/src/chains/ethereum/options/node_modules
lerna info clean removing /home/david/work/ganache-core/src/chains/ethereum/utils/node_modules
lerna info clean removing /home/david/work/ganache-core/src/chains/tezos/options/node_modules
lerna info clean removing /home/david/work/ganache-core/src/chains/tezos/tezos/node_modules
lerna info clean removing /home/david/work/ganache-core/src/packages/cli/node_modules
lerna info clean removing /home/david/work/ganache-core/src/packages/core/node_modules
lerna info clean removing /home/david/work/ganache-core/src/packages/flavors/node_modules
lerna info clean removing /home/david/work/ganache-core/src/packages/ganache/node_modules
lerna info clean removing /home/david/work/ganache-core/src/packages/options/node_modules
lerna info clean removing /home/david/work/ganache-core/src/packages/promise-queue/node_modules
lerna info clean removing /home/david/work/ganache-core/src/packages/utils/node_modules
lerna success clean finished

> root@ tsc.clean /home/david/work/ganache-core
> npx lerna exec -- npx shx rm -rf lib dist

npx: installed 766 in 31.316s
lerna notice cli v3.22.1
lerna info versioning independent
lerna info Executing command in 12 packages: "npx shx rm -rf lib dist"
npx: installed 21 in 3.513sact:shx: verb lock using /home/david/.npm/_locks/staging-0393b40e8a8e169d.lock for /home/david/.npm/_npx/936938/lib/node_modules/.staging
npx: installed 21 in 3.599sesh-package-json:shx: timing action:finalize Completed in 21ms
npx: installed 21 in 3.643s
npx: installed 21 in 3.54s
npx: installed 21 in 3.545s
npx: installed 21 in 3.464s
npx: installed 21 in 3.585s
npx: installed 21 in 3.268s
npx: installed 21 in 3.468s
npx: installed 21 in 3.529s
npx: installed 21 in 3.39s
npx: installed 21 in 3.337s
lerna success exec Executed command in 12 packages: "npx shx rm -rf lib dist"
```

mikeseese commented on src/chains/ethereum/ethereum/src/blockchain.ts:
```diff
@@ -256,7 +256,7 @@ export default class Blockchain extends Emittery.Typed<
         if (options.chain.time != null) {
           // If we were given a timestamp, use it instead of the `_currentTime`
-          const t = +options.chain.time;
+          const t = options.chain.time.getTime();
```
I changed this to `.getTime()` because I changed the type of `chain.time` to `Date` instead of `Date | number`. I'm not sure why there was a `| number`. It's not in either legacy documentation that a user can provide a UTC millisecond timestamp; both CLI and Core documentation states it should be a ISO 8601 string or a Date object (core only)

I also double checked, and the `ganache-core@2` does assume it's a Date:
- https://github.com/trufflesuite/ganache-core/blob/develop/lib/blockchain_double.js#L88-L90
- https://github.com/trufflesuite/ganache-core/blob/develop/lib/blockchain_double.js#L1251-L1255

davidmurdoch replied:
good catch

davidmurdoch commented on src/chains/ethereum/options/src/chain-options.ts:
```diff
@@ -134,36 +136,66 @@ export type ChainConfig = {
+    legacyName: "hardfork",
+    cliAliases: ["k", "hardfork"],
+    cliType: "string",
+    cliChoices: Object.keys(enumerate<Hardfork>())
```
nice.

davidmurdoch commented on src/chains/ethereum/options/src/chain-options.ts:
```diff
@@ -134,36 +136,66 @@ export type ChainConfig = {
+    defaultDescription:
+      "System time at process start or Network ID of forked blockchain if configured.",
+    legacyName: "network_id",
+    cliAliases: ["i", "networkId", "network_id"],
```
`"network_id"` shouldn't be a CLI alias, as it doesn't work for `ganache-cli@latest` as it is today.

mikeseese replied:
fair, I partially was considering removing the gap of the confusion between core and cli. i'm fine with removing it though 👍

davidmurdoch commented on src/chains/ethereum/options/src/logging-options.ts:
```diff
@@ -71,17 +82,32 @@ const logger: Logger = { log: () => {} };
   logger: {
     normalize,
+    shortDescription:
+      " An object, like `console`, that implements a `log` function.",
```
```suggestion
      "An object, like `console`, that implements a `log` function.",
```

davidmurdoch commented on src/chains/ethereum/options/src/miner-options.ts:
```diff
@@ -136,39 +141,80 @@ export type MinerConfig = {
   callGasLimit: {
     normalize: Quantity.from,
+    shortDescription:
+      "Sets the transaction gas limit in WEI for `eth_call` and eth_estimateGas` calls.",
```
```suggestion
      "Sets the transaction gas limit in WEI for `eth_call` and `eth_estimateGas` calls.",
```

davidmurdoch commented on src/chains/ethereum/options/src/wallet-options.ts:
```diff
@@ -181,45 +190,106 @@ export type WalletConfig = {
-    default: () => randomAlphaNumericString(10, alea()),
-    legacyName: "seed"
+    shortDescription: "Seed to use to generate a mnemonic.",
+    default: config =>
```
I just realized that my `default(config)` implementation depends on the order the options are specified in _this_ Definitions object.

The users options will be parsed in the order they appear in `Object.keys(WalletOptions)`.

Which would result in `seed` being parsed before `deterministic`, ensuring deterministic is _never_ set to `true`.

I haven't tested, but if my hunch here is true, the easiest thing to do is to just move `deterministic` right before `seed` in this `WalletOptions` object. And slap a big ol' comment in here about how the ordering is required (and test ensuring the correct wallet options get applied would be a nice bonus :-) ).

There are other more "technically correct" ways of solving this, but all the ones I've thought of have their own downsides and limitations (like maintaining dependency trees or a double pass over the options).

mikeseese replied:
Looking at the code in `packages/options/src/create.ts`, I concur with your hunch 👍 Thanks for the heads up!

mikeseese replied:
I also manually tested and confirmed that the order does matter (your hunch is correct)

davidmurdoch reviewed (commented):
This is just a partial review with some comments. I'll do another pass at it later.

mikeseese committed:
Apply suggestions from code review

Co-authored-by: David Murdoch <davidmurdoch@users.noreply.github.com>

mikeseese committed:
remove network_id as a cli alias

mikeseese committed:
make sure deterministic option is before seed

davidmurdoch committed:
Merge branch 'next' into feat/dynamic-cli

davidmurdoch committed:
fix: remove ts-transformer-enumerate so running without compiling works

davidmurdoch committed:
add cli start command to root

davidmurdoch committed:
fix some package.json `"test"` `directories`

davidmurdoch committed:
fix some package.json homepage & repository links

davidmurdoch committed:
stop vscode from auto attaching to completion.sh's use of node

davidmurdoch committed:
ensure DefaultOptionsByName is fully typed

davidmurdoch committed:
update the minimum supported dev node version

davidmurdoch committed:
update typescript to v4.1.3

davidmurdoch committed:
update prettier to v 2.2.1

davidmurdoch committed:
remove default_balance_ether from cli aliases as it was never a cli option in v2

davidmurdoch committed:
rename cli-related settings from `serverSettings` to `cliSettings`

davidmurdoch committed:
remove keepAliveTimeout from server settings

davidmurdoch committed:
ensure we can log before exit when needed

also fixes crash when already exitting when the server couldn't close, and we tried to call `setBlocking`, but on the wrong object!

davidmurdoch committed:
fix main package webpack after recent directory restructure

davidmurdoch committed:
add root start script to start the cli

davidmurdoch committed:
remove unsued files

davidmurdoch committed:
clarify ganache compatibility
